tokio-quiche/src/http3/driver/mod.rs

```diff
@@ -577,15 +625,36 @@ impl<H: DriverHooks> H3Driver<H> {
             h3::Event::Finished => self.process_h3_fin(qconn, stream_id),
 
             h3::Event::Reset(code) => {
-                if let Some(ctx) = self.stream_map.get(&stream_id) {
-                    ctx.audit_stats.set_recvd_reset_stream_error_code(code as _);
-                }
+                if let Some(ctx) = self.stream_map.get_mut(&stream_id) {
+                    ctx.handle_recvd_reset(code);
+                    // See if we are waiting on this stream and close the channel
+                    // if we are. If we are not waiting, `handle_recvd_reset()`
+                    // will have taken care of closing.
+                    for pending in self.waiting_streams.iter_mut() {
+                        match pending {
+                            WaitForStream::Upstream(
+                                WaitForUpstreamCapacity {
+                                    stream_id: id,
+                                    chan: Some(chan),
+                                },
+                            ) if stream_id == *id => {
+                                chan.close();
+                            },
+                            _ => {},
+                        }
+                    }
 
-                self.h3_event_sender
-                    .send(H3Event::ResetStream { stream_id }.into())
```

Comment:
I think this may actually be a bug in the current implementation. If we're acting as a server and receive a RST_STREAM, that just means the client isn't going to write anymore, but by sending the `ResetStream` event upstream we'll kill the request handler and then never respond. This is similar to what Nginx does, but I don't think it's technically correct per the RFC.

I think, in the client-side case, it's fine - receiving a RST_STREAM from the server would mean that it's never going to send anything after that, so killing the request future is okay. cc @LPardue not sure if we want to change this now or as follow-up

Reply:
Cool, agreed. Approved